lib/simple_token_authentication.ex

```diff
@@ -40,17 +40,39 @@ defmodule SimpleTokenAuthentication do
     end
   end
 
-  defp matches?(token, value) when is_binary(token) and is_binary(value),
-    do: String.trim(token) != "" && secure_compare(token, value)
-    
-  defp matches?(token, value) when is_tuple(token) and is_binary(value) do
-    if matches?(elem(token, 1), value) do
-      Logger.metadata(service_name: elem(token, 0))
-      true
-    else
-      false
-    end
+  def build_token_map() do
+    tokens =
+      :simple_token_authentication
+      |> Application.get_env(:token)
+      |> List.wrap()
+
+    token_map = build_map_from_token_list(tokens, :global)
+
+    services =
+      :simple_token_authentication
+      |> Application.get_env(:service_tokens)
+      |> List.wrap()
+
+    token_map =
+      Enum.reduce(services, token_map, fn {name, tokens}, acc ->
+        service_map = build_map_from_token_list(List.wrap(tokens), name)
+        Map.merge(acc, service_map)
+      end)
+
+    :persistent_term.put(:simple_token_authentication_map, token_map)
+
+    token_map
   end
 
-  defp matches?(_, _), do: false
+  def build_map_from_token_list(tokens, service_name) do
```

Comment:
If we go this way, would it make more sense to reduce into the same map instead of merging several smaller maps?
Also with a case at the top level of a function, one could optionally just match with the function if preferred for clarity.
```elixir
defp build_map_from_token_list(tokens, service_name, token_map) do
  Enum.reduce(tokens, token_map, fn
    "", acc ->
      acc

    token, acc ->
      Map.put(acc, token, service_name)
  end)
end
```
Though I'm a fan of simplifying with comprehensions, when possible (as in clearer)
```elixir
defp build_token_map do
  global =
    :simple_token_authentication
    |> Application.get_env(:token)
    |> List.wrap()

  services =
    :simple_token_authentication
    |> Application.get_env(:service_tokens)
    |> List.wrap()

  token_map =
    for {service, tokens} <- [{:global, global} | services],
        token <- tokens,
        token != "",
        reduce: %{} do
      acc ->
        Map.put(acc, token, service)
    end

  # ...
end
```

Reply:
yea I had a feeling there was a better way but could not come up with it. Will fix it to use comprehension as it will make this more readable